### PR TITLE
Add update dyanmo functionality

### PR DESCRIPTION
**Pull request by Calum-Campbell**
- Adds the functionality for the dynamo dependancy to update a switch remotely 
- Also adds functionality to "get" the state of the switches live and surpass the scheduler
- Bumps version to 0.4 (still needs to be published)

Used in :
- https://github.com/guardian/discussion-platform/pull/277
 - https://github.com/guardian/discussion-api/pull/557

### CODE REVIEW

File: src/main/scala/com/gu/dynamodbswitches/Switch.scala [hunk 1]
```diff
@@ -1,9 +1,20 @@
...
+
 case class Switch(name: String, default: Boolean = false) {
   @volatile private var state: Boolean = default
 
   def enabled: Boolean = state
 
   def enabled_=(value: Boolean) = state = value
+
+  def toStringAttribute() = {
+    new AttributeValue().withS(this.name)
```
Comment (TBonnin) [inline src/main/scala/com/gu/dynamodbswitches/Switch.scala]: Can you not store boolean in DynamoDB?

File: src/main/scala/com/gu/dynamodbswitches/Switches.scala [hunk 2]
```diff
@@ -35,6 +34,27 @@ trait Switches extends Logging {
         error(s"Encountered Amazon service error when trying to update switches from DynamoDB: ${exception.getMessage}")
     }
   }
+
+    def updateDynamo(switches : List[Switch]) = {
+    val listWR = switches.map(switch => {
+      new WriteRequest(new PutRequest(Map("name" -> switch.toStringAttribute(), "enabled" -> switch.toNumAttributeValue()).asJava))
+    }).asJava
```
Comment (TBonnin) [inline src/main/scala/com/gu/dynamodbswitches/Switches.scala]: No need for the inside `{}`

File: src/main/scala/com/gu/dynamodbswitches/Switches.scala [hunk 3]
```diff
@@ -35,6 +34,27 @@ trait Switches extends Logging {
...
+      case e: Exception => println(e)
+      case _ => println("Something went wrong")
+    }
+  }
+
+  def getSwitchboardState(): List[Switch] = {
+    val results = dynamoDbClient.scan(new ScanRequest(dynamoDbTableName)).getItems.asScala.toList.map(_.asScala.toMap)
+    results.map(item => {
+      Switch(item("name").getS, item("enabled").getN == "1")
+    })
```
Comment (TBonnin) [inline src/main/scala/com/gu/dynamodbswitches/Switches.scala]: No need for the inside `{}`
You also don't need the intermediary val `results` and the double `map`. What about:
```
dynamoDbClient
  .scan(new ScanRequest(dynamoDbTableName))
  .getItems.asScala.toList
  .map { 
    val item = _.asScala.toMap
    Switch(item("name").getS, item("enabled").getN == "1")
  }
```

File: src/main/scala/com/gu/dynamodbswitches/Switch.scala [hunk 4]
```diff
@@ -1,9 +1,20 @@
...
+  }
+
+  def toNumAttributeValue() = {
```
Comment (nicl) [inline src/main/scala/com/gu/dynamodbswitches/Switch.scala]: This seems like an odd method. It converts the default value to a string containing 1 or 0.

Perhaps a better name might be `defaultAsAttributeValue`? 

But even better might be to inline this logic, or define a more generic and pure method in a companion object. E.g.

    def asAttributeValue(b: Boolean) = {
      val str = if(b) "1" else "0"
      new AttributeValue().withN(str)
    }

File: src/main/scala/com/gu/dynamodbswitches/Switches.scala [hunk 5]
```diff
@@ -35,6 +34,27 @@ trait Switches extends Logging {
...
+    def updateDynamo(switches : List[Switch]) = {
+    val listWR = switches.map(switch => {
+      new WriteRequest(new PutRequest(Map("name" -> switch.toStringAttribute(), "enabled" -> switch.toNumAttributeValue()).asJava))
+    }).asJava
+
+    try {
+      dynamoDbClient.batchWriteItem(Map(dynamoDbTableName -> listWR).asJava)
+    } catch {
+      case e: Exception => println(e)
+      case _ => println("Something went wrong")
```
Comment (nicl) [inline src/main/scala/com/gu/dynamodbswitches/Switches.scala]: Ideally we'd log this, but I can see adding support for a logger might be easier in a separate PR.

Comment (Calum-Campbell) [inline src/main/scala/com/gu/dynamodbswitches/Switches.scala]: Yep changed to error and warn for now?

**Pull request merged**